dhcpv4/dhcpv4.go

```diff
@@ -416,6 +416,52 @@ func (o Options) GetOneOption(code OptionCode) Option {
 	return nil
 }
 
+// Marshal writes options binary representations to b.
+func (o Options) Marshal(b *uio.Lexer, writeEnd bool) {
+	for _, opt := range o {
+		code := opt.Code()
+
+		// Even if the End option is in there, don't marshal it until
+		// the end.
+		if code == OptionEnd {
```

Comment:
I'll move this to a different PR, but yes, and early End should always be considered an anomaly.

The other side should never send you an early End. I mean, you wouldn't even *know* if it's an early End. You have no idea if the data after the End is garbage or not.

And a user of this library should never be given the opportunity to include an early End in the options list, because it's invalid anyway. A user of this library shouldn't be able to insert the End option at all, I'd assert, and the library itself should take care of that while marshaling.

Reply:
Your reasoning is valid, but there is another one equally valid in my opinion - this library was initially inspired by `scapy`'s concepts, where you can intentionally craft malformed packets for security testing purposes, if you want to. Then sane (hopefully) defaults have been implemented for the regular user, and validation methods have been added for the paranoid